File: helfomat-api/src/main/java/de/helfenkannjeder/helfomat/api/organization/event/OrganizationEventAssembler.java
package de.helfenkannjeder.helfomat.api.organization.event;

import de.helfenkannjeder.helfomat.api.organization.OrganizationAssembler;
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor;
import de.helfenkannjeder.helfomat.core.organization.event.ConfirmedChangeOrganizationEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationCreateEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationDeleteEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddAddressEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddAttendanceTimeEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddContactPersonEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddGroupEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddPictureEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddQuestionAnswerEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditAddVolunteerEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDefaultAddressEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteAddressEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteAttendanceTimeEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteContactPersonEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteGroupEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeletePictureEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteQuestionAnswerEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDeleteVolunteerEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditDescriptionEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditLogoEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditNameEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditTeaserImageEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditUrlNameEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEditWebsiteEvent;
import de.helfenkannjeder.helfomat.core.organization.event.OrganizationEvent;
import de.helfenkannjeder.helfomat.core.organization.event.ProposedChangeOrganizationEvent;
import de.helfenkannjeder.helfomat.core.question.Question;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Valentin Zickner
 */
public class OrganizationEventAssembler implements OrganizationEventVisitor<OrganizationEventDto> {

    private final List<Question> questions;

    public OrganizationEventAssembler(List<Question> questions) {
        this.questions = questions;
    }

    public static List<OrganizationEventDto> toOrganizationEventDto(List<OrganizationEvent> organizationEvents, List<Question> questions) {
        OrganizationEventAssembler organizationEventAssembler = new OrganizationEventAssembler(questions);
        return organizationEvents.stream()
            .map(organizationEvent -> organizationEvent.visit(organizationEventAssembler))
            .collect(Collectors.toList());
    }

    @Override
    public OrganizationEventDto visit(OrganizationCreateEvent organizationCreateEvent) {
        return new OrganizationCreateEventDto(
            organizationCreateEvent.getOrganizationId(),
            organizationCreateEvent.getName(),
            organizationCreateEvent.getUrlName(),
            organizationCreateEvent.getOrganizationType()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationDeleteEvent organizationDeleteEvent) {
        return new OrganizationDeleteEventDto(organizationDeleteEvent.getOrganizationId());
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddAddressEvent organizationEditAddAddressEvent) {
        return new OrganizationEditAddAddressEventDto(
            organizationEditAddAddressEvent.getOrganizationId(),
            organizationEditAddAddressEvent.getIndex(),
            OrganizationAssembler.toAddressDto(organizationEditAddAddressEvent.getAddress())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddAttendanceTimeEvent organizationEditAddAttendanceTimeEvent) {
        return new OrganizationEditAddAttendanceTimeEventDto(
            organizationEditAddAttendanceTimeEvent.getOrganizationId(),
            organizationEditAddAttendanceTimeEvent.getIndex(),
            OrganizationAssembler.toAttendanceTimeDto(organizationEditAddAttendanceTimeEvent.getAttendanceTime())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddContactPersonEvent organizationEditAddContactPersonEvent) {
        return new OrganizationEditAddContactPersonEventDto(
            organizationEditAddContactPersonEvent.getOrganizationId(),
            organizationEditAddContactPersonEvent.getIndex(),
            OrganizationAssembler.toContactPersonDto(organizationEditAddContactPersonEvent.getContactPerson())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddGroupEvent organizationEditAddGroupEvent) {
        return new OrganizationEditAddGroupEventDto(
            organizationEditAddGroupEvent.getOrganizationId(),
            organizationEditAddGroupEvent.getIndex(),
            OrganizationAssembler.toGroupDto(organizationEditAddGroupEvent.getGroup())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddPictureEvent organizationEditAddPictureEvent) {
        return new OrganizationEditAddPictureEventDto(
            organizationEditAddPictureEvent.getOrganizationId(),
            organizationEditAddPictureEvent.getIndex(),
            organizationEditAddPictureEvent.getPictureId()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddQuestionAnswerEvent organizationEditAddQuestionAnswerEvent) {
        return new OrganizationEditAddQuestionAnswerEventDto(
            organizationEditAddQuestionAnswerEvent.getOrganizationId(),
            organizationEditAddQuestionAnswerEvent.getIndex(),
            OrganizationAssembler.toAnsweredQuestionDto(
                organizationEditAddQuestionAnswerEvent.getQuestionAnswer(),
                OrganizationAssembler.determineQuestionText(questions, organizationEditAddQuestionAnswerEvent.getQuestionAnswer().getQuestionId())
            )
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditAddVolunteerEvent organizationEditAddVolunteerEvent) {
        return new OrganizationEditAddVolunteerEventDto(
            organizationEditAddVolunteerEvent.getOrganizationId(),
            organizationEditAddVolunteerEvent.getIndex(),
            OrganizationAssembler.toVolunteerDto(organizationEditAddVolunteerEvent.getVolunteer())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDefaultAddressEvent organizationEditDefaultAddressEvent) {
        return new OrganizationEditDefaultAddressEventDto(
            organizationEditDefaultAddressEvent.getOrganizationId(),
            OrganizationAssembler.toAddressDto(organizationEditDefaultAddressEvent.getDefaultAddress())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteAddressEvent organizationEditDeleteAddressEvent) {
        return new OrganizationEditDeleteAddressEventDto(
            organizationEditDeleteAddressEvent.getOrganizationId(),
            OrganizationAssembler.toAddressDto(organizationEditDeleteAddressEvent.getAddress())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteAttendanceTimeEvent organizationEditDeleteAttendanceTimeEvent) {
        return new OrganizationEditDeleteAttendanceTimeEventDto(
            organizationEditDeleteAttendanceTimeEvent.getOrganizationId(),
            OrganizationAssembler.toAttendanceTimeDto(organizationEditDeleteAttendanceTimeEvent.getAttendanceTime())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteContactPersonEvent organizationEditDeleteContactPersonEvent) {
        return new OrganizationEditDeleteContactPersonEventDto(
            organizationEditDeleteContactPersonEvent.getOrganizationId(),
            OrganizationAssembler.toContactPersonDto(organizationEditDeleteContactPersonEvent.getContactPerson())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteGroupEvent organizationEditDeleteGroupEvent) {
        return new OrganizationEditDeleteGroupEventDto(
            organizationEditDeleteGroupEvent.getOrganizationId(),
            OrganizationAssembler.toGroupDto(organizationEditDeleteGroupEvent.getGroup())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeletePictureEvent organizationEditDeletePictureEvent) {
        return new OrganizationEditDeletePictureEventDto(
            organizationEditDeletePictureEvent.getOrganizationId(),
            organizationEditDeletePictureEvent.getPictureId()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteQuestionAnswerEvent organizationEditDeleteQuestionAnswerEvent) {
        return new OrganizationEditDeleteQuestionAnswerEventDto(
            organizationEditDeleteQuestionAnswerEvent.getOrganizationId(),
            OrganizationAssembler.toAnsweredQuestionDto(
                organizationEditDeleteQuestionAnswerEvent.getQuestionAnswer(),
                OrganizationAssembler.determineQuestionText(questions, organizationEditDeleteQuestionAnswerEvent.getQuestionAnswer().getQuestionId())
            )
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDeleteVolunteerEvent organizationEditDeleteVolunteerEvent) {
        return new OrganizationEditDeleteVolunteerEventDto(
            organizationEditDeleteVolunteerEvent.getOrganizationId(),
            OrganizationAssembler.toVolunteerDto(organizationEditDeleteVolunteerEvent.getVolunteer())
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditDescriptionEvent organizationEditDescriptionEvent) {
        return new OrganizationEditDescriptionEventDto(
            organizationEditDescriptionEvent.getOrganizationId(),
            organizationEditDescriptionEvent.getDescription()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditLogoEvent organizationEditLogoEvent) {
        return new OrganizationEditLogoEventDto(
            organizationEditLogoEvent.getOrganizationId(),
            organizationEditLogoEvent.getLogo()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditNameEvent organizationEditNameEvent) {
        return new OrganizationEditNameEventDto(
            organizationEditNameEvent.getOrganizationId(),
            organizationEditNameEvent.getName()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditTeaserImageEvent organizationEditTeaserImageEvent) {
        return new OrganizationEditTeaserImageEventDto(
            organizationEditTeaserImageEvent.getOrganizationId(),
            organizationEditTeaserImageEvent.getTeaserImage()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditUrlNameEvent organizationEditUrlNameEvent) {
        return new OrganizationEditUrlNameEventDto(
            organizationEditUrlNameEvent.getOrganizationId(),
            organizationEditUrlNameEvent.getUrlName()
        );
    }

    @Override
    public OrganizationEventDto visit(OrganizationEditWebsiteEvent organizationEditWebsiteEvent) {
        return new OrganizationEditWebsiteEventDto(
            organizationEditWebsiteEvent.getOrganizationId(),
            organizationEditWebsiteEvent.getWebsite()
        );
    }

    @Override
    public OrganizationEventDto visit(ProposedChangeOrganizationEvent proposedChangeOrganizationEvent) {
        return new ProposedChangeOrganizationEventDto(
            proposedChangeOrganizationEvent.getOrganizationId(),
            proposedChangeOrganizationEvent.getAuthor(),
            proposedChangeOrganizationEvent.getSources(),
            proposedChangeOrganizationEvent.getChanges()
                .stream()
                .map(organizationEvent -> organizationEvent.visit(this))
                .collect(Collectors.toList())
        );
    }

    @Override
    public OrganizationEventDto visit(ConfirmedChangeOrganizationEvent confirmedChangeOrganizationEvent) {
        return new ConfirmedChangeOrganizationEventDto(
            confirmedChangeOrganizationEvent.getOrganizationId(),
            confirmedChangeOrganizationEvent.getApprovedBy(),
            confirmedChangeOrganizationEvent.getAuthor(),
            confirmedChangeOrganizationEvent.getSources(),
            confirmedChangeOrganizationEvent.getChanges()
                .stream()
                .map(organizationEvent -> organizationEvent.visit(this))
                .collect(Collectors.toList())
        );
    }
}

File: helfomat-api/src/main/kotlin/de/helfenkannjeder/helfomat/api/organization/event/OrganizationEventAssembler.kt
package de.helfenkannjeder.helfomat.api.organization.event

import de.helfenkannjeder.helfomat.api.organization.*
import de.helfenkannjeder.helfomat.core.organization.Organization
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor
import de.helfenkannjeder.helfomat.core.organization.event.*
import de.helfenkannjeder.helfomat.core.question.Question

/**
 * @author Valentin Zickner
 */
class OrganizationEventAssembler(
    private val questions: List<Question>,
    originalOrganization: Organization?
) : OrganizationEventVisitor<OrganizationEventDto> {

    private var currentOrganization: Organization? = originalOrganization?.copy()

    override fun visit(organizationCreateEvent: OrganizationCreateEvent) =
        OrganizationCreateEventDto(organizationCreateEvent.organizationId, organizationCreateEvent.name, organizationCreateEvent.urlName, organizationCreateEvent.organizationType, isEventApplicable(organizationCreateEvent))

    override fun visit(organizationDeleteEvent: OrganizationDeleteEvent) =
        OrganizationDeleteEventDto(organizationDeleteEvent.organizationId, isEventApplicable(organizationDeleteEvent))

    override fun visit(organizationEditAddAddressEvent: OrganizationEditAddAddressEvent) =
        OrganizationEditAddAddressEventDto(organizationEditAddAddressEvent.organizationId, organizationEditAddAddressEvent.index, organizationEditAddAddressEvent.address.toAddressDto(), isEventApplicable(organizationEditAddAddressEvent))

    override fun visit(organizationEditAddAttendanceTimeEvent: OrganizationEditAddAttendanceTimeEvent) =
        OrganizationEditAddAttendanceTimeEventDto(organizationEditAddAttendanceTimeEvent.organizationId, organizationEditAddAttendanceTimeEvent.index, organizationEditAddAttendanceTimeEvent.attendanceTime.toAttendanceTimeDto(), isEventApplicable(organizationEditAddAttendanceTimeEvent))

    override fun visit(organizationEditAddContactPersonEvent: OrganizationEditAddContactPersonEvent) =
        OrganizationEditAddContactPersonEventDto(organizationEditAddContactPersonEvent.organizationId, organizationEditAddContactPersonEvent.index, organizationEditAddContactPersonEvent.contactPerson.toContactPersonDto(), isEventApplicable(organizationEditAddContactPersonEvent))

    override fun visit(organizationEditAddGroupEvent: OrganizationEditAddGroupEvent) =
        OrganizationEditAddGroupEventDto(organizationEditAddGroupEvent.organizationId, organizationEditAddGroupEvent.index, organizationEditAddGroupEvent.group.toGroupDto(), isEventApplicable(organizationEditAddGroupEvent))

    override fun visit(organizationEditAddPictureEvent: OrganizationEditAddPictureEvent) =
        OrganizationEditAddPictureEventDto(organizationEditAddPictureEvent.organizationId, organizationEditAddPictureEvent.index, organizationEditAddPictureEvent.pictureId, isEventApplicable(organizationEditAddPictureEvent))

    override fun visit(organizationEditAddQuestionAnswerEvent: OrganizationEditAddQuestionAnswerEvent) =
        OrganizationEditAddQuestionAnswerEventDto(
            organizationEditAddQuestionAnswerEvent.organizationId,
            organizationEditAddQuestionAnswerEvent.index,
            organizationEditAddQuestionAnswerEvent.questionAnswer.toAnsweredQuestionDto(questions.getAnswerToQuestion(organizationEditAddQuestionAnswerEvent.questionAnswer.questionId)),
            isEventApplicable(organizationEditAddQuestionAnswerEvent)
        )

    override fun visit(organizationEditAddVolunteerEvent: OrganizationEditAddVolunteerEvent) =
        OrganizationEditAddVolunteerEventDto(organizationEditAddVolunteerEvent.organizationId, organizationEditAddVolunteerEvent.index, organizationEditAddVolunteerEvent.volunteer.toVolunteerDto(), isEventApplicable(organizationEditAddVolunteerEvent))

    override fun visit(organizationEditDefaultAddressEvent: OrganizationEditDefaultAddressEvent) =
        OrganizationEditDefaultAddressEventDto(organizationEditDefaultAddressEvent.organizationId, organizationEditDefaultAddressEvent.defaultAddress?.toAddressDto(), isEventApplicable(organizationEditDefaultAddressEvent))

    override fun visit(organizationEditDeleteAddressEvent: OrganizationEditDeleteAddressEvent) =
        OrganizationEditDeleteAddressEventDto(organizationEditDeleteAddressEvent.organizationId, organizationEditDeleteAddressEvent.address.toAddressDto(), isEventApplicable(organizationEditDeleteAddressEvent))

    override fun visit(organizationEditDeleteAttendanceTimeEvent: OrganizationEditDeleteAttendanceTimeEvent) =
        OrganizationEditDeleteAttendanceTimeEventDto(organizationEditDeleteAttendanceTimeEvent.organizationId, organizationEditDeleteAttendanceTimeEvent.attendanceTime.toAttendanceTimeDto(), isEventApplicable(organizationEditDeleteAttendanceTimeEvent))

    override fun visit(organizationEditDeleteContactPersonEvent: OrganizationEditDeleteContactPersonEvent) =
        OrganizationEditDeleteContactPersonEventDto(organizationEditDeleteContactPersonEvent.organizationId, organizationEditDeleteContactPersonEvent.contactPerson.toContactPersonDto(), isEventApplicable(organizationEditDeleteContactPersonEvent))

    override fun visit(organizationEditDeleteGroupEvent: OrganizationEditDeleteGroupEvent) =
        OrganizationEditDeleteGroupEventDto(organizationEditDeleteGroupEvent.organizationId, organizationEditDeleteGroupEvent.group.toGroupDto(), isEventApplicable(organizationEditDeleteGroupEvent))

    override fun visit(organizationEditDeletePictureEvent: OrganizationEditDeletePictureEvent) =
        OrganizationEditDeletePictureEventDto(organizationEditDeletePictureEvent.organizationId, organizationEditDeletePictureEvent.pictureId, isEventApplicable(organizationEditDeletePictureEvent))

    override fun visit(organizationEditDeleteQuestionAnswerEvent: OrganizationEditDeleteQuestionAnswerEvent) =
        OrganizationEditDeleteQuestionAnswerEventDto(
            organizationEditDeleteQuestionAnswerEvent.organizationId,
            organizationEditDeleteQuestionAnswerEvent.questionAnswer.toAnsweredQuestionDto(questions.getAnswerToQuestion(organizationEditDeleteQuestionAnswerEvent.questionAnswer.questionId)),
            isEventApplicable(organizationEditDeleteQuestionAnswerEvent)
        )

    override fun visit(organizationEditDeleteVolunteerEvent: OrganizationEditDeleteVolunteerEvent) =
        OrganizationEditDeleteVolunteerEventDto(organizationEditDeleteVolunteerEvent.organizationId, organizationEditDeleteVolunteerEvent.volunteer.toVolunteerDto(), isEventApplicable(organizationEditDeleteVolunteerEvent))

    override fun visit(organizationEditDescriptionEvent: OrganizationEditDescriptionEvent) =
        OrganizationEditDescriptionEventDto(organizationEditDescriptionEvent.organizationId, organizationEditDescriptionEvent.description, isEventApplicable(organizationEditDescriptionEvent))

    override fun visit(organizationEditLogoEvent: OrganizationEditLogoEvent) =
        OrganizationEditLogoEventDto(organizationEditLogoEvent.organizationId, organizationEditLogoEvent.logo, isEventApplicable(organizationEditLogoEvent))

    override fun visit(organizationEditNameEvent: OrganizationEditNameEvent) =
        OrganizationEditNameEventDto(organizationEditNameEvent.organizationId, organizationEditNameEvent.name, isEventApplicable(organizationEditNameEvent))

    override fun visit(organizationEditTeaserImageEvent: OrganizationEditTeaserImageEvent) =
        OrganizationEditTeaserImageEventDto(organizationEditTeaserImageEvent.organizationId, organizationEditTeaserImageEvent.teaserImage, isEventApplicable(organizationEditTeaserImageEvent))

    override fun visit(organizationEditUrlNameEvent: OrganizationEditUrlNameEvent) =
        OrganizationEditUrlNameEventDto(organizationEditUrlNameEvent.organizationId, organizationEditUrlNameEvent.urlName, isEventApplicable(organizationEditUrlNameEvent))

    override fun visit(organizationEditWebsiteEvent: OrganizationEditWebsiteEvent) =
        OrganizationEditWebsiteEventDto(organizationEditWebsiteEvent.organizationId, organizationEditWebsiteEvent.website, isEventApplicable(organizationEditWebsiteEvent))

    override fun visit(proposedChangeOrganizationEvent: ProposedChangeOrganizationEvent) =
        ProposedChangeOrganizationEventDto(
            proposedChangeOrganizationEvent.organizationId,
            proposedChangeOrganizationEvent.author,
            proposedChangeOrganizationEvent.sources,
            proposedChangeOrganizationEvent.changes.map { it.visit(this) },
            true
        )

    override fun visit(confirmedChangeOrganizationEvent: ConfirmedChangeOrganizationEvent) =
        ConfirmedChangeOrganizationEventDto(
            confirmedChangeOrganizationEvent.organizationId,
            confirmedChangeOrganizationEvent.approvedBy,
            confirmedChangeOrganizationEvent.author,
            confirmedChangeOrganizationEvent.sources,
            confirmedChangeOrganizationEvent.changes.map { it.visit(this) },
            true
        )

    override fun visit(organizationEditChangeGroupEvent: OrganizationEditChangeGroupEvent): OrganizationEventDto =
        OrganizationEditChangeGroupEventDto(
            organizationEditChangeGroupEvent.organizationId,
            organizationEditChangeGroupEvent.indexOffset,
            organizationEditChangeGroupEvent.oldGroup.toGroupDto(),
            organizationEditChangeGroupEvent.group.toGroupDto(),
            isEventApplicable(organizationEditChangeGroupEvent)
        )

    override fun visit(organizationEditChangePictureEvent: OrganizationEditChangePictureEvent): OrganizationEventDto =
        OrganizationEditChangePictureEventDto(
            organizationEditChangePictureEvent.organizationId,
            organizationEditChangePictureEvent.indexOffset,
            organizationEditChangePictureEvent.pictureId,
            isEventApplicable(organizationEditChangePictureEvent)
        )

    override fun visit(organizationEditChangeAttendanceTimeEvent: OrganizationEditChangeAttendanceTimeEvent): OrganizationEventDto =
        OrganizationEditChangeAttendanceTimeEventDto(
            organizationEditChangeAttendanceTimeEvent.organizationId,
            organizationEditChangeAttendanceTimeEvent.indexOffset,
            organizationEditChangeAttendanceTimeEvent.oldAttendanceTime.toAttendanceTimeDto(),
            organizationEditChangeAttendanceTimeEvent.attendanceTime.toAttendanceTimeDto(),
            isEventApplicable(organizationEditChangeAttendanceTimeEvent)
        )

    override fun visit(organizationEditChangeAddressEvent: OrganizationEditChangeAddressEvent): OrganizationEventDto =
        OrganizationEditChangeAddressEventDto(
            organizationEditChangeAddressEvent.organizationId,
            organizationEditChangeAddressEvent.indexOffset,
            organizationEditChangeAddressEvent.oldAddress.toAddressDto(),
            organizationEditChangeAddressEvent.address.toAddressDto(),
            isEventApplicable(organizationEditChangeAddressEvent)
        )

    override fun visit(organizationEditChangeQuestionAnswerEvent: OrganizationEditChangeQuestionAnswerEvent): OrganizationEventDto =
        OrganizationEditChangeQuestionAnswerEventDto(
            organizationEditChangeQuestionAnswerEvent.organizationId,
            organizationEditChangeQuestionAnswerEvent.indexOffset,
            organizationEditChangeQuestionAnswerEvent.oldQuestionAnswer.toAnsweredQuestionDto(questions.getAnswerToQuestion(organizationEditChangeQuestionAnswerEvent.oldQuestionAnswer.questionId)),
            organizationEditChangeQuestionAnswerEvent.questionAnswer.toAnsweredQuestionDto(questions.getAnswerToQuestion(organizationEditChangeQuestionAnswerEvent.questionAnswer.questionId)),
            isEventApplicable(organizationEditChangeQuestionAnswerEvent)
        )

    override fun visit(organizationEditChangeContactPersonEvent: OrganizationEditChangeContactPersonEvent): OrganizationEventDto =
        OrganizationEditChangeContactPersonEventDto(
            organizationEditChangeContactPersonEvent.organizationId,
            organizationEditChangeContactPersonEvent.indexOffset,
            organizationEditChangeContactPersonEvent.oldContactPerson.toContactPersonDto(),
            organizationEditChangeContactPersonEvent.contactPerson.toContactPersonDto(),
            isEventApplicable(organizationEditChangeContactPersonEvent)
        )

    private fun isEventApplicable(organizationEvent: OrganizationEvent): Boolean {
        val organization = currentOrganization ?: return true
        val builder = Organization.Builder(organization)
        organizationEvent.applyOnOrganizationBuilder(builder)
        val newOrganization = builder.build()
        val isApplicable = organization != newOrganization
        currentOrganization = newOrganization
        return isApplicable
    }
}

fun OrganizationEvent.toOrganizationEventDto(questions: List<Question>, originalOrganization: Organization?) = this.toOrganizationEventDto(OrganizationEventAssembler(questions, originalOrganization))
fun OrganizationEvent.toOrganizationEventDto(organizationEventAssembler: OrganizationEventAssembler) = this.visit(organizationEventAssembler)
fun List<OrganizationEvent>.toOrganizationEventDtos(questions: List<Question>, originalOrganization: Organization?): List<OrganizationEventDto> {
    val organizationEventAssembler = OrganizationEventAssembler(questions, originalOrganization)
    return this.map { it.toOrganizationEventDto(organizationEventAssembler) }
}